### ISSUE TITLE
Parsing/indexing previously unopened files for large project seem to be not working 

### ISSUE THREAD

**Issue by csbaroti** [microsoft/vscode-cpptools#7332]
**Type: LanguageService**

**Describe the bug**
- OS and Version: SLES12-SP2-12.2
- VS Code Version: 1.50 / 1.54.2
- C/C++ Extension Version: 1.1.3 / 1.2.2
- Other extensions you installed (and if the issue persists after disabling them): disabled all, still exists
- Does this issue involve using SSH remote to run the extension on a remote machine?: no

Single project workspace, project size, ~36k entries in compile_commands.json
I am now trying to test out vscode for c++ development on our rather large project and some features fail.

Hover symbol lookup, autocompletion seem to work fine, however "Find All References", "Switch Header / Source"
do not work (does not find the files) when the files have not been previously open in the editor, if they were open even in previous, now closed window, then these features usually work.  (When i first open a file, it has its includes with blue underline then searching for include paths and resolving all of them just fine in like ~5-10s)

Thus my question, is this a bug / config issue or is this lazy parsing/indexing is by design?
If it is by design, can there be a feature request to allow pre-parsing/indexing?

The IDE we use currently does indexing and parsing in few minutes so that previously mentioned basic features work even in files outside of workspace, specified in compile_commands.json

I tried to look for same kind of issues here, in them they used browse.path, i used it too with few key dirs but does not solve my issue. Could you help out? Otherwise like this extension very much and waiting for other features.
In logs below, had to remove file paths as they contain sensitive information.

**Steps to reproduce**

1. Go to any cc/hh file that has not been opened yet
2. Click on "Find All References" or "Switch Header / Source"
3. Wait until finishes
4. See that only previously parsed/index files are there, others not

**Expected behavior**

Use above mentioned features as if the source code files have been open in the editor because right now it's the only way

<details>
  <summary><strong>Logs</strong></summary>

```
settings.json
{
    "search.useIgnoreFiles": false,
    "search.useGlobalIgnoreFiles": true,
    "C_Cpp.intelliSenseCachePath": "${workspaceFolder}/build/intellisense.cache",
    "C_Cpp.workspaceParsingPriority": "highest",
    "C_Cpp.default.browse.databaseFilename": "${workspaceFolder}/build/vscode.db",
    "C_Cpp.loggingLevel": "Debug",
    "files.associations": {
        "string": "cpp"
    },
    "files.exclude": {
        "**/.git": true,
        "**/.svn": true,
        "**/.hg": true,
        "**/CVS": true,
        "**/.DS_Store": true,
        "**/*Doxy*": true,
        "**/*bin": true,
        "**/*obj": true,
        "**/tools": true,
        "**/?*.[^jcdhx]*": true,
        "**/*.c[^c]*": true,
        "**/*.h[^h]*": true,
		... others...
    },
    "search.exclude": {
        "**/node_modules": true,
        "**/bower_components": true,
        "**/*.code-search": true,
        "**/.git": true,
        "**/.svn": true,
        "**/.hg": true,
        "**/CVS": true,
        "**/.DS_Store": true,
        "**/*Doxy*": true,
        "**/*bin": true,
        "**/*obj": true,
        "**/tools": true,
        "**/testing": true,
        "**/opt": true,
        "**/repo": true,
        "**/build": true,
		... others...
    }
}

c_cpp_properties.json
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
		... other includes....
                "${workspaceFolder}/..long_path.../**",
            ],
            "compilerPath": "${workspaceFolder}/..long_path.../x86_64-dx-linux-gnu-g++",
            "cStandard": "c11",
            "cppStandard": "c++11",
            "compileCommands": "${workspaceFolder}/compile_commands.json",
            "intelliSenseMode": "gcc-x64",
            "configurationProvider": "ms-vscode.cmake-tools",
            "browse": {
                "databaseFilename": "${workspaceFolder}/build/vscode_browse.db",
                "path": [
                    "${workspaceFolder}",
                    ... other paths to source code under workspace...
                ]
            }
        }
    ],
    "version": 4
}

[DEBUG] the language server when opening previously not opened file
textDocument/didOpen: file://B.hh
Checking for syntax errors: file://B.hh
Queueing IntelliSense update for files in translation unit of: /A.cc
cpptools/activeDocumentChange: file://B.hh
cpptools/textEditorSelectionChange
cpptools/getDocumentSymbols: file://B.hh (id: 49)
cpptools/getSemanticTokens: file://B.hh (id: 50)
cpptools/getFoldingRanges: file://B.hh (id: 51)
cpptools/getDocumentSymbols
  tag parsing file: B.hh
cpptools/getCodeActions: file://B.hh (id: 52)
idle loop: reparsing the active document
Checking for syntax errors: file://B.hh
Queueing IntelliSense update for files in translation unit of: /A.cc
cpptools/finishUpdateSquiggles
Error squiggle count: 0
Error squiggle count: 0
terminating child process: 7882
Update IntelliSense time (sec): 2.155
cpptools/getSemanticTokens: file:///A.cc (id: 53)
cpptools/getCodeActions: file://B.hh (id: 54)
cpptools/getFoldingRanges: file://B.hh (id: 55)

WHEN running Find All References this message in the output:
[Warning] Some references may be missing, because workspace parsing was incomplete when Find All References was started.


```
</details>

**Screenshots**

**Additional context**

As the project is huge, after opening vscode "Unable to watch for files changes" appears.
The database sign remains "Discovering files: 0".

**Comment by sean-mcmanus** [comment]
Your situation doesn't sound normal. If you run C/C++: Log Diagnostics it should show a message near the end about what Browse.path is being used based on the compile commands. Your files.excludes setting might be causing the files to be skipped (if you remove those are the files parsed?). If you set C_Cpp.loggingLevel to "Debug" you should be able to see "tag parsing" messages for the files that it is processing after you run the C/C++: Reset IntelliSense Database command.

**Comment by ebarcsa** [comment]
Okay, I think I managed to debug my problem thanks to your instructions.

Previously I thought it wasn't discovering my files, actually it was the opposite, discovered too many (500k+)
because I manually added includes which were added to the compile_commands.json,
because our other IDE had it included and I thought vscode won't use includePath if compileCommands is given.
Well there was one directory containing all modules with lots of folders and files underneath.

Conclusion i had to remove that directory so database icon could actually show discovery progress with percent sign.
Took couple minutes and mentioned features work great. 448MB .db 369MB intellisense cache right now.

I could also mention that when it was i think discovering or Find Reference, maybe both, when i closed VSCode cpptools process was still in the background and had to killall.

Thanks for helping, Hopefully, there will be time for features like "Call Hierarchy" or finding derived classes / implementation more easily, basically code navigation, still very usable now.
